### PR TITLE
The enforcer rule to entry point JARs for BOMs correctly

### LINKED ISSUE #1932: The enforcer rule does not properly count entry point JARs for BOM project
While I was trying to use the enforcer rule for the shared-dependencies BOM (https://github.com/googleapis/java-shared-dependencies/issues/269), I found that the Linkage Checker enforcer rule does not detect the dependency conflict of grpc-census 1.35.0 v.s. opencensus-api 0.28.1 when reportOnlyReachable is true.

I found the problem is that the enforcer rule does not count the direct dependencies correctly when the BOM project imports another BOM.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1932 , where the enforcer rule was not detecting problems when reportOnlyReachable is true.

### CODE REVIEW

File: enforcer-rules/src/it/bom-project-no-error/pom.xml [hunk 1]
```diff
@@ -35,16 +35,12 @@
...
+      unreachable. https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1871 -->
       <dependency>
-        <groupId>com.google.api-client</groupId>
-        <artifactId>google-api-client</artifactId>
-        <version>1.27.0</version>
-      </dependency>
-      <dependency>
-        <groupId>io.grpc</groupId>
-        <artifactId>grpc-core</artifactId>
-        <version>1.17.1</version>
```
Comment (suztomo) [inline enforcer-rules/src/it/bom-project-no-error/pom.xml:47]: This test case was incorrect. The linkage errors on gRPC core was reachable:

```
[INFO] Building: bom-project-no-error/pom.xml
[INFO] run post-build script verify.groovy
[INFO]   The post-build script did not succeed. assert buildLog.text.contains("No reachable error found")
       |        |    |
       |        |    false
       |        Listening for transport dt_socket at address: 5005
       |        [INFO] Scanning for projects...
       |        [INFO]
       |        [INFO] -------< com.google.cloud.tools.opensource:bom-project-no-error >-------
       |        [INFO] Building bom-project-no-error 1.0-SNAPSHOT
       |        [INFO] --------------------------------[ pom ]---------------------------------
       |        [INFO]
       |        [INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-linkage-checker) @ bom-project-no-error ---
       |        [ERROR] Linkage Checker rule found 1 reachable error:
       |        (com.google.guava:guava:20.0) com.google.common.base.Verify's method verify(boolean, String, Object) is not found;
       |          referenced by 3 class files
       |            io.grpc.internal.ServiceConfigInterceptor (io.grpc:grpc-core:1.17.1)
       |            io.grpc.internal.JndiResourceResolverFactory (io.grpc:grpc-core:1.17.1)
       |            io.grpc.internal.DnsNameResolver (io.grpc:grpc-core:1.17.1)
       |          Cause:
       |            Dependency conflict: com.google.guava:guava:20.0 does not define com.google.common.base.Verify's method verify(boolean, String, Object) but com.google.guava:guava:26.0-android defines it.
       |              selected: com.google.api-client:google-api-client:1.27.0 (compile) / com.google.guava:guava:20.0 (compile)
       |              unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
       |            Dependency conflict: com.google.guava:guava:20.0 does not define com.google.common.base.Verify's method verify(boolean, String, Object) but com.google.guava:guava:26.0-android defines it.
       |              selected: com.google.api-client:google-api-client:1.27.0 (compile) / com.google.guava:guava:20.0 (compile)
       |              unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
       |            Dependency conflict: com.google.guava:guava:20.0 does not define com.google.common.base.Verify's method verify(boolean, String, Object) but com.google.guava:guava:26.0-android defines it.
       |              selected: com.google.api-client:google-api-client:1.27.0 (compile) / com.google.guava:guava:20.0 (compile)
       |              unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
       |        Problematic artifacts in the dependency tree:
       |        com.google.guava:guava:20.0 is at:
       |          com.google.api-client:google-api-client:1.27.0 (compile) / com.google.guava:guava:20.0 (compile)
       |          and 3 other dependency paths.
       |        io.grpc:grpc-core:1.17.1 is at:
       |          io.grpc:grpc-core:1.17.1 (compile)
       |
```

File: enforcer-rules/src/it/bom-project-no-error/pom.xml [hunk 2]
```diff
@@ -35,16 +35,12 @@
...
+        <groupId>com.google.http-client</groupId>
+        <artifactId>google-http-client</artifactId>
```
Comment (suztomo) [inline enforcer-rules/src/it/bom-project-no-error/pom.xml:42]: The dependencies of this is artifact have unreachable linkage errors. Details: https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1871

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 3]
```diff
@@ -192,12 +192,7 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
         return;
       }
 
-      // As sorted by level order, the first elements in classpath are the project and its direct
-      // non-test dependencies.
-      List<org.apache.maven.model.Dependency> dependencies = project.getDependencies();
-      long projectDependencyCount =
-          dependencies.stream().filter(dependency -> !"test".equals(dependency.getScope())).count();
-      List<ClassPathEntry> entryPoints = classPath.subList(0, (int) projectDependencyCount + 1);
```
Comment (suztomo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java:200]: This logic did not take dependencyManagement into account.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 4]
```diff
@@ -381,4 +378,29 @@ private ClassPathResult findBomClasspath(
...
+   * section.
+   */
+  private ImmutableList<ClassPathEntry> entryPoints(
+      MavenProject project, ImmutableList<ClassPathEntry> classPath) {
+    // As sorted by level order, the first elements in classpath are the project and its direct
+    // non-test dependencies.
+    List<org.apache.maven.model.Dependency> directDependencies =
+        dependencySection == DependencySection.DEPENDENCY_MANAGEMENT
+            ? project.getDependencyManagement().getDependencies()
+            : project.getDependencies();
```
Comment (suztomo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java:398]: This is the fix. The entry point JARs are the direct dependencies for non-BOM project, and the ones in dependencyManagement sections for BOM project.

File: enforcer-rules/src/it/report-only-reachable-bom/pom.xml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        <artifactId>grpc-bom</artifactId>
+        <version>1.35.0</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+      <dependency>
+        <groupId>io.opencensus</groupId>
+        <artifactId>opencensus-api</artifactId>
+        <version>0.28.1</version>
+      </dependency>
```
Comment (suztomo) [inline enforcer-rules/src/it/report-only-reachable-bom/pom.xml:53]: Test case for a BOM that imports another BOM (such as the shared-dependencies BOM).

File: enforcer-rules/src/it/report-only-reachable-bom/pom.xml [hunk 6]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ Copyright 2020 Google LLC.
```
Comment (elharo) [inline enforcer-rules/src/it/report-only-reachable-bom/pom.xml:4]: 2021

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 7]
```diff
@@ -381,4 +378,29 @@ private ClassPathResult findBomClasspath(
     }
     return result;
   }
+
+  /**
+   * Returns the class path entries that contains entry point classes.
```
Comment (elharo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java:383]: contains --> contain

File: enforcer-rules/src/it/report-only-reachable-bom/pom.xml (hunk 6 above)
Comment (suztomo) [reply enforcer-rules/src/it/report-only-reachable-bom/pom.xml]: Done

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java (hunk 7 above)
Comment (suztomo) [reply enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java]: Fixed.

**Review by suztomo — COMMENTED** [review]
@elharo PTAL

**Comment by suztomo** [comment]
cla/google ?

**Review by elharo — COMMENTED** [review]
need to merge in master

**Review by elharo — COMMENTED** [review]
Why is windows build failing?

**Review by elharo — COMMENTED** [review]
Test failure might be related:



[ERROR] Failures: 


[ERROR]   ClassPathBuilderTest.testBomToPaths_firstElementsAreBomMembers:102 expected: com.google.api:gax-httpjson:0.57.0

but was : com.google.code.findbugs:jsr305:3.0.2

**Comment by suztomo** [comment]
artifact download failure might cause that. Triggered rebuild.

**Review by elharo — APPROVED** [review]

**Pull request merged**